pkg/jx/cmd/step/syntax/step_syntax_canonical.go

```diff
@@ -0,0 +1,414 @@
+package syntax
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/jenkins-x/jx/pkg/config"
+	"github.com/jenkins-x/jx/pkg/gits"
+	"github.com/jenkins-x/jx/pkg/jenkinsfile"
+	"github.com/jenkins-x/jx/pkg/jenkinsfile/gitresolver"
+	"github.com/jenkins-x/jx/pkg/jx/cmd/helper"
+	"github.com/jenkins-x/jx/pkg/jx/cmd/opts"
+	"github.com/jenkins-x/jx/pkg/kube"
+	"github.com/jenkins-x/jx/pkg/log"
+	"github.com/jenkins-x/jx/pkg/tekton/syntax"
+	"github.com/jenkins-x/jx/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
+)
+
+// StepSyntaxCanonicalOptions contains the command line flags
+type StepSyntaxCanonicalOptions struct {
+	opts.StepOptions
+
+	Pack              string
+	BuildPackURL      string
+	BuildPackRef      string
+	Context           string
+	CustomImage       string
+	DefaultImage      string
+	NoKaniko          bool
+	KanikoImage       string
+	ProjectID         string
+	DockerRegistry    string
+	DockerRegistryOrg string
+	SourceName        string
+	CustomEnvs        []string
+
+	PodTemplates map[string]*corev1.Pod
+
+	GitInfo         *gits.GitRepository
+	VersionResolver *opts.VersionResolver
+}
+
+// NewCmdStepSyntaxCanonical Creates a new Command object
+func NewCmdStepSyntaxCanonical(commonOpts *opts.CommonOptions) *cobra.Command {
+	options := &StepSyntaxCanonicalOptions{
+		StepOptions: opts.StepOptions{
+			CommonOptions: commonOpts,
+		},
+	}
+
+	cmd := &cobra.Command{
+		Use:   "canonical",
+		Short: "Outputs a canonical representation of the pipeline to be executed",
+		Long:  "Reads the appropriate jenkins-x.yml, depending on context, from the current directory, if one exists, and outputs a canonical representation of the pipelines",
+		Run: func(cmd *cobra.Command, args []string) {
+			options.Cmd = cmd
+			options.Args = args
+			err := options.Run()
+			helper.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().StringArrayVarP(&options.CustomEnvs, "env", "e", nil, "List of custom environment variables to be applied to resources that are created")
+
+	options.AddCommonFlags(cmd)
+	return cmd
+}
+
+// AddCommonFlags adds common CLI options
+func (o *StepSyntaxCanonicalOptions) AddCommonFlags(cmd *cobra.Command) {
```

Comment:
Would we not need some options to control the output? Thinking about how I would like to plug this into the meta pipeline generation, I'd like to be able to write this somehow to file. 

Not sure whether it should override/create an existing `jenkins-x.yml` or whether per default it should create something like `jenkins-x-context-effective.yml`

Printing to stdout seems to be more of a dev/dry-run kind of thing.

Reply:
The easiest in terms of pipeline extensions and Apps trying to contribute to the pipeline would probably be to create/modify the appropriate jenkins-x-<context>.yml

Reply:
Yeah, that's why this is WIP. =)

Reply:
Take a look at my latest change - adds `--output-dir` and `--output-file` options. If neither's given, it'll just go to stdout. If `--output-dir` is given and `--output-file` is not, it'll write a file named `jenkins-x[-context]-canonical.yml` to that directory. If `--output-file` is given, that'll be used instead of `jenkins-x[-context]-canonical.yml`, and if `--output-file` is given and `--output-dir` is *not*, then that file will be written to the current directory.

I'm 99.99% sure this could be more clear. =)

Reply:
> Take a look at my latest change

Will do. Some thoughts prior to actually having tried this out.

> - adds --output-dir and --output-file options. If neither given, it'll just go to stdout. 

I am usually a bit skeptical when there is an option for specifying an output directory and file. It tends to get ambiguous. What happens for example if output-dir is specified, but output-file is a path? If it is a relative path, do you create the according directory structure under output-dir or is this an error? What if output-file is an absolute path?

To a certain degree --output-dir seems to be a flag in disguise which really mainly decides whether to write to stdout or to disk. --output-file alone would be enough to get the output to where one wants, either by transparently creating intermediate directories or by returning an error if any of the intermediate directories does not exit. 

Here is another idea. Why not write per default to disk to a default file and have a --dry-run to print to stdout. --dry-run aligns more with other commands. I like the idea to establish some common patterns. The --output-file lets you then choose a different name and path for the output. 

> it'll write a file named jenkins-x[-context]-canonical.yml to that directory. 

Initially I was leaning towards writing  (potentially overwriting) jenkins-x[-context].yml, but the idea of creating this canonical yaml is growing on me. Overwriting leaves a bad aftertaste. 

So the contract for the meta-pipeline is then, that an extension gets passed the context and from that can infer the pipeline config it should modify. Alternatively it gets passed the full path to the canonical config.

What's about in the meta pipeline construction we set something like JX_PIPELINE_CONTEXT and JX_PIPELINE_CANONICAL_CONFIG as environment variables for all steps following the one calling `jx step syntax canonical`.

Reply:
A final thought. Is canonical a synonym for effective? I am coming from the Java/Maven world were one talks about the effective POM. For me canonical is a bit different to effective, but I am not sure.

Reply:
I am completely open to a better option/config setup here - I literally just threw something out as fast as I could. =) And yes, "canonical" is a synonym for "effective" here. I also come from the Maven world and that's why I was referring to it as "effective pipeline" initially, but I'm just not sure if it's the right term. I'd rather go with that, though - "canonical" is feeling worse to me as time goes on.

Reply:
My vote would go for 'effective'. I connect 'canonical' with normalization which is not really what's happening here. What we do here is generating the net pipeline config. But if you feel strongly about 'canonical' that's fine for me as well.